File: backend/app/schemas/article.py
from datetime import datetime
from enum import Enum

from pydantic import BaseModel, ConfigDict, Field


class CategoryEnum(str, Enum):
    AI_ML = "AI/ML"
    BACKEND = "Backend"
    FRONTEND = "Frontend"
    DEVOPS = "DevOps"


class ArticleBase(BaseModel):
    title: str = Field(..., max_length=500)
    content: str
    author: str = Field(..., max_length=100)
    category: CategoryEnum
    published_at: datetime


class ArticleCreate(ArticleBase):
    """Schema for creating an article."""

    pass


class ArticleUpdate(BaseModel):
    """Schema for updating an article (partial update)."""

    title: str | None = Field(None, max_length=500)
    content: str | None = None
    author: str | None = Field(None, max_length=100)
    category: CategoryEnum | None = None
    published_at: datetime | None = None


class ArticleResponse(ArticleBase):
    """Schema for article detail response (includes full content)."""

    model_config = ConfigDict(from_attributes=True)

    id: int
    created_at: datetime
    updated_at: datetime


class ArticleSummary(BaseModel):
    """Schema for article list item (excerpt instead of full content)."""

    model_config = ConfigDict(from_attributes=True)

    id: int
    title: str
    excerpt: str
    author: str
    category: CategoryEnum
    published_at: datetime
    created_at: datetime
    updated_at: datetime

File: frontend/src/types/article.ts
export type Category = 'AI/ML' | 'Backend' | 'Frontend' | 'DevOps';

export interface Article {
  id: number;
  title: string;
  content: string;
  author: string;
  category: Category;
  published_at: string;
  created_at: string;
  updated_at: string;
}

export interface ArticleSummary {
  id: number;
  title: string;
  excerpt: string;
  author: string;
  category: Category;
  published_at: string;
  created_at: string;
  updated_at: string;
  highlight?: string;
}

export interface SearchResult {
  id: number;
  title: string;
  excerpt: string;
  author: string;
  category: Category;
  published_at: string;
  similarity: number;
  highlight?: string;
}

export interface ArticleCreateInput {
  title: string;
  content: string;
  author: string;
  category: Category;
  published_at: string;
}

export interface ArticleUpdateInput {
  title?: string;
  content?: string;
  author?: string;
  category?: Category;
  published_at?: string;
}

export interface Pagination {
  total: number;
  page: number;
  page_size: number;
  total_pages: number;
}
